### ISSUE TITLE
Shortest query runs out memory

### ISSUE THREAD

**Issue by chenyuzhi459** [dgraph-io/dgraph#2816]
- What version of Dgraph are you using?
v1.0.10.

- Have you tried reproducing the issue with latest release?
No.

- What is the hardware spec (RAM, OS)?
 64G, CentOS7

- Steps to reproduce the issue (command/config used to run Dgraph).
1. Make shortest path query with  param `numpaths:2`, for expamle:
```
{
      path as shortest(from: 0x24a, to: 0x4,numpaths: 2) {
        ~purchase
        purchase
      }
      path(func: uid(path)) {
        uid
      }
}
```
And query result is:
```
{
  "data": {
    "path": [
      {
        "uid": "0x24a"
      },
      {
        "uid": "0x29"
      },
      {
        "uid": "0x4e3"
      },
      {
        "uid": "0x4"
      }
    ],
    "_path_": [
      {
        "purchase": [
          {
            "~purchase": [
              {
                "purchase": [
                  {
                    "uid": "0x4"
                  }
                ],
                "uid": "0x4e3"
              }
            ],
            "uid": "0x29"
          }
        ],
        "uid": "0x24a"
      },
      {
        "purchase": [
          {
            "~purchase": [
              {
                "purchase": [
                  {
                    "uid": "0x4"
                  }
                ],
                "uid": "0x14b0"
              }
            ],
            "uid": "0x43"
          }
        ],
        "uid": "0x24a"
      }
    ]
  }
}
```
Obviously, the depth of this shortest-path is 2. 

2. Add the param `depth:1`  for the first query:
```
{
      path as shortest(from: 0x24a, to: 0x4, numpaths: 2, detph:1) {
        ~purchase
        purchase
      }
      path(func: uid(path)) {
        uid
      }
}
```
Then the alpha server  consumes memory until it runs out all of memory(64G) in the background. And finally it's killed by system.

----

However, if I remove the param `numpaths` in the second query:
```
{
      path as shortest(from: 0x24a, to: 0x4, detph:1) {
        ~purchase
        purchase
      }
      path(func: uid(path)) {
        uid
      }
}
```
it seems return the correct result:
```
Your query did not return any results
```

- Expected behaviour and actual result.

Expected return result is `Your query did not return any results`, just like the third query result.  But  it's blocked all the time without returning the result actually.

- Auxiliary information

Alpha server config(cluster mode):
```
lru_mb: 6144
query_edge_limit : 100000000
```
System process info:
There are  some other java programs that don't take up much memory  like zoopkeeper, hdfs run on the server.

**Comment by manishrjain** [comment]
Seems like `numpaths` and `depth` args are interacting in a weird way. @codexnull can you have a look?

**Comment by martinmr** [comment]
I have been looking at the shortest path code in #2915. I think the issue is that the k-shortest path algorithm used to allow loops. I have removed that since once the minimum and maximum weight parameters are introduced looping will not make much sense.

It's possible that the algorithm gets stuck exploring a loop in the graph and dgraph eventually runs out of memory. Not sure how to reproduce since the OP didn't share the data they are using but once the PR above is merged into master it would make sense to check if the issue is still there.

**Comment by campoy** [comment]
Hey @martinmr , could you check whether this issue is still occurring and update it accordingly?

Thanks!

**Comment by martinmr** [comment]
This was fixed by the PR mentioned above but I forgot to close this issue.

**Comment by BlankRain** [comment]
@martinmr  hi martin ,which PR fixed this issue? could you give me the link ?

**Comment by martinmr** [comment]
@BlankRain it should be https://github.com/dgraph-io/dgraph/pull/2915

**Comment by BlankRain** [comment]
@martinmr thank you
